src/pudl/transform/vceregen.py

```diff
@@ -0,0 +1,257 @@
+    purposes. The report_year column is also helpful for filtering,
+    so we keep all three!
+
+    For leap years (2020), December 31st is excluded.
+    """
+    logger.info("Adding time columns")
+    # This data is compiled for modeling purposes and skips the last
+    # day of a leap year. When adding a datetime column, we need
+    # to make sure that we skip the 31st of December, 2020 and that
+    # every year has exactly 8760 hours in it.
+    all_years = df.report_year.unique()
+    datetime8760_index = pd.DatetimeIndex(
+        pd.concat(
+            [
+                pd.Series(pd.date_range(start=f"{year}-01-01", periods=8760, freq="h"))
+                for year in all_years
+            ]
+        )
+    )
+    new_time_col = pd.DataFrame(
+        {
+            "hour_utc": datetime8760_index,
+        }
+    )
+    df = pd.concat(
+        [df.reset_index(drop=True), new_time_col.reset_index(drop=True)], axis=1
+    ).rename(columns={"unnamed_0": "hour_of_year"})
+    # Make sure that leapyear date doesn't exist
+    if not df[df["hour_utc"] == "2020-12-31 01:00:00"].empty:
+        raise AssertionError("There should be no Dec-31 in 2020")
+    # Validate that the datetime aligns with the hour of the year
+    df["hour_from_date"] = (
+        df["hour_utc"].dt.hour + (df["hour_utc"].dt.dayofyear - 1) * 24 + 1
+    )
+    if not df[df["hour_from_date"] != df["hour_of_year"]].empty:
+        raise AssertionError("datetime columns doesn't align with hour of year")
+    df = df.drop(columns="hour_from_date")
+    return df
+
+
+def _stack_cap_fac_df(df: pd.DataFrame, df_name: str) -> pd.DataFrame:
+    """Funciton to transform each capacity factor table individually to save memory.
+
+    The main transforms are turning county columns into county rows and renaming columns
+    to be more human-readable and compatible with the FIPs df that will get merged in.
+
+    This function is intended to save memory by being applied to each individual
+    capacity factor table rather than the giant combined one.
+    """
+    logger.info("Stacking the county columns")
+    df_stacked = (
+        df.set_index(["hour_utc", "hour_of_year", "report_year"])
+        .stack()
+        .reset_index()
+        .rename(
+            columns={"level_3": "county_state_names", 0: f"capacity_factor_{df_name}"}
+        )
+        .assign(county_state_names=lambda x: pd.Categorical(x.county_state_names))
+    )
+    return df_stacked
+
+
+def _make_cap_fac_frac(df: pd.DataFrame, df_name: str) -> pd.DataFrame:
+    """Make the capacity factor column a fraction instead of a percentage.
+
+    This step happens before the table gets stacked to save memory.
+    """
+    logger.info("Converting capacity factor into a fraction.")
+    county_cols = [x for x in df.columns if x not in ["report_year", "unnamed_0"]]
+    df[county_cols] = df[county_cols] / 100
+    if (max_val := df[county_cols].max().max()) > 1.02:
```

Comment:
I'm not exactly sure why, but there are some capacity factor values in the solar table that are ~102. This isn't so far over 100 that I'm concerned, but I'm curious. Is there any explanation for this?

Reply:
Chis said this is due to cooler panel temperatures.